src/docs/arch-block.dox

```diff
@@ -1,9 +1,208 @@
+When disabled the block manager still does perform a checksum operation but only
+the first 64 bytes of the buffer are included.
+
+The checksum is used when reading blocks to validate their contents, it
+is compared with the checksum extracted from the address cookie and it is
+compared with a checksum generated from the buffer that was held in the block
+being read. In both cases the checksum should match.
+
+@section block_usage How WiredTiger uses the block manager
+
+@subsection creation File creation and the block manager
+
+When a new file is created in WiredTiger via WT_SESSION::create, the file is
+created on disk and the associated \c allocation_size is written out to the
+metadata file. However the block manager itself isn't created. Only when a btree
+is opened does WiredTiger allocate a block manager. Each btree has its own
+block manager.
+
+@subsection read Reading files and pages
+
+When an existing btree is opened for the first time, the location of the root
+block is contained in the metadata file \c WiredTiger.wt. The block manager will
+read the block at the location specified and return the page image as a buffer
+to the layer above. This will then be instantiated as a page in memory.
+
+From there subsequent page addresses can be read from the root page and the
+process repeated as required. If a cursor traverses to a page which hasn't been
+read into memory the same process will take place.
+
+@subsection Writing
+
+Two cases exist for writing out data using the block manager: checkpoint and
+eviction. When a page image is written out the block manager the \c bm->write
+API is called. See \c bt_io.c for more detail.
+
+@subsubsection Checkpoint
+
+For details on checkpoint at the WiredTiger level see @ref arch-checkpoint.
+
+At the block manager level, a checkpoint corresponds with a set of blocks that
+are stored in the file associated with the given URI. Typically each file will
+contain a minimum of two checkpoints. Upon opening an existing file the most
+recent checkpoint is read.
+
+During a checkpoint new blocks are only written out for dirty pages. A block can
+be included in multiple checkpoints. Assuming a page \c X is dirty and gets
+checkpointed in checkpoint \c A, it will be created as a new block on disk. Now
+the same page \c X isn't modified and another checkpoint is taken. The page is
+clean and as such will not require a new block to be written for it. The address
+of the original block is still valid.
+
+Checkpoints are created in depth first order, leaf blocks are created, then
+the parent blocks. This is a requirement as the parent blocks contain the
+addresses of the leaf blocks.
+
+The block manager doesn't guarantee that calling \c bm->write will result in
+the data being flushed to disk. In the checkpoint scenario WiredTiger will also
+call \c bm->sync which will call the file system dependent flush function.
+
+As a checkpoint progresses it updates its three extent lists, these extent lists
```

Comment:
I fixed up the sentence structure to make the paragraph more readable. Although you might be rewriting this section anyway? 

_As a checkpoint progresses it updates its three extent lists. These extent lists
are written out to disk in blocks as part of the checkpoint. When a new checkpoint
is written, the previous checkpoints become candidates for deletion. For example, suppose checkpoints \c A and \c B exist on disk and the block manager just finished writing a new checkpoint \c C. The block manager only retains two checkpoints by default, so it can delete checkpoint \c A. Checkpoint \c A's extent lists get merged into checkpoint \c B's extent lists. If block \c N is created in checkpoint \c A and removed in checkpoint \c B, then the space occupied by the block is considered free and can be utilized._

Reply:
Reworded: 2627290